### PR TITLE
Fixed issue #17188: {GROUPNAME} and other variables not rendered

### PR DESCRIPTION
**Pull request by gabrieljenik**
Adding EM processing to group description

### CODE REVIEW

File: application/helpers/SurveyRuntimeHelper.php [hunk 1]
```diff
@@ -433,7 +433,7 @@ public function run($surveyid, $args)
 
             $aGroup['showgroupinfo'] = $showgroupinfo_;
             $aGroup['showdescription']  = (!$this->previewquestion && trim($gl['description']) != "" && $showgroupdesc_);
-            $aGroup['description']      = $gl['description'];
+            $aGroup['description']      = LimeExpressionManager::ProcessString($gl['description'], null, null, 3, 1);
```
Comment (Shnoulle) [inline application/helpers/SurveyRuntimeHelper.php:436]: We process 2 times then ? 

Seems GROUPNAME have an issue, but process description is another issue (and i don't think we need it).

Comment (Shnoulle) [reply application/helpers/SurveyRuntimeHelper.php:436]: https://github.com/LimeSurvey/LimeSurvey/blob/435badbd8f103673ece463a6c2bea8a4296ff25f/themes/survey/vanilla/views/subviews/survey/group_subviews/group_desc.twig#L25

**Review by Shnoulle — COMMENTED** [review]
I don't lime to process string 2 times. And if some theme manager use it differently ? It broke.

File: application/helpers/SurveyRuntimeHelper.php (hunk 1 above)
Comment (c-schmitz) [inline application/helpers/SurveyRuntimeHelper.php:436]: I agree.

**Review by c-schmitz — CHANGES_REQUESTED** [review]

**Comment by Shnoulle** [comment]
I update the mantis : https://bugs.limesurvey.org/view.php?id=17188

With some test  on 3.25.19 : GID is OK, always, GROUNAME is OK in group by group, not in all in one.

File: application/helpers/SurveyRuntimeHelper.php (hunk 1 above)
Comment (gabrieljenik) [reply application/helpers/SurveyRuntimeHelper.php:436]: Yes, I agree.

**Comment by gabrieljenik** [comment]
I think GID and Groupname are not rendered because of this.
 https://github.com/gabrieljenik/LimeSurvey/blob/62000f0f910761b305c39ec7e2b0875bb2c16f68/application/helpers/SurveyRuntimeHelper.php#L513

Any idea why is that there?

**Comment by Shnoulle** [comment]
1. GID is rendered in 3.X
2. GID and GROUPNAME is rendered in Grou by group mode

GID and GROUPNAME must be set to empty when we are not inside a group.
I think it was added for this a long time ago.

Capture d’écran_2021-03-26_14-07-07

**Comment by gabrieljenik** [comment]
> GID and GROUPNAME must be set to empty when we are not inside a group.

Maybe we are setting it this all time? That could be the issue?

**Comment by Shnoulle** [comment]
See my screenshot : issue **only** in all in one page.

**Comment by gabrieljenik** [comment]
But that's for LS3 or LS4?

**Comment by Shnoulle** [comment]
See mantis : 4.X OK on group by group mode

**Comment by gabrieljenik** [comment]
We will debug, thanks

**Comment by Shnoulle** [comment]
> Any idea why is that there?

If you remove it : in 4.X, you have GID of group 2 in group 1

Capture d’écran_2021-03-27_09-08-30

**Comment by Shnoulle** [comment]
The only solution to get in in groupdescription : it's your solution  … 
https://github.com/LimeSurvey/LimeSurvey/pull/1821/files#diff-f2c0610f506dc0397d2efd8d8203a78dc44f2825e26eb0ac70efd87205d2e36eR436

I don't see other way.

Must be done in GROUPNAME too ? if you use for title of group : Group {GID} ?

**Comment by gabrieljenik** [comment]
I would just remove the EM Processing for the group on the twig.

Afetr some research we found that when going in "all in one" mode, twig only has EM as setup initially. 
Later when looping through the groups, EM instance is not refreshed, so I think there is the issue. Also, I don't think EM has the info from all groups loaded.

So removing processString from twig could be good.
The issue is the backward compatibility.

Another thought, going deeper it may be conceptually wrong that twig process the EM for that.
Does twig (a presentation layer) has all the info needed to call the EM?

**Comment by olleharstedt** [comment]
Hm.

**Comment by maziminke** [comment]
"processString()" is used at themes quite often so let's not change that or make sure we guarantee backward compatiblitly.

**Comment by Shnoulle** [comment]
> "processString()" is used at themes quite often so let's not change that or make sure we guarantee backward compatiblitly.

I make some test : even with the fix of @gabrieljenik : there are no issue to processtring again.

A solution : update group_desc.twig `{{processString(aGroup.description) }}`

by
```
            {{ processString(
                aGroup.description,
                0,
                3,
                ({
                    'GROUPNAME' : aGroup.name,
                    'QID' : '',
                    'GID' : aGroup.gid
                })
            ) }}
```

It work in 3.X except aGroup.gid is empty

@gabrieljenik maybe you can fix it like this ? @maziminke you can update your template now :)

**Comment by gabrieljenik** [comment]
Hi All,

I believe a vaible solution oving forward:
- Keep my fix
- removing processString from twig, as moving forward.
- As for backward compatibility shouldn't be an issue. Themes that double process, shouldn't throw errors according to Denis.

Will check if GroupName accepts EM in LS2

**Comment by gabrieljenik** [comment]
> > "processString()" is used at themes quite often so let's not change that or make sure we guarantee backward compatiblitly.
> 
> I make some test : even with the fix of @gabrieljenik : there are no issue to processtring again.
> 
> A solution : update group_desc.twig `{{processString(aGroup.description) }}`
> 
> by
> 
> ```
>             {{ processString(
>                 aGroup.description,
>                 0,
>                 3,
>                 ({
>                     'GROUPNAME' : aGroup.name,
>                     'QID' : '',
>                     'GID' : aGroup.gid
>                 })
>             ) }}
> ```
> 
> It work in 3.X except aGroup.gid is empty
> 
> @gabrieljenik maybe you can fix it like this ? @maziminke you can update your template now :)

Why would you reprocess with all that information, which could be faulty sometimes at twig level (case All in one)?

**Comment by Shnoulle** [comment]
> 
> Why would you reprocess with all that information, which could be faulty sometimes at twig level (case All in one)?


I didn't reprocess any information : i use https://github.com/LimeSurvey/LimeSurvey/blob/6437c998731e1e79da24c394ef205444cfa75cdf/application/helpers/expressions/em_manager_helper.php#L4143 

I sent the array. I have group.name, but not group.gid in 3.X

Adding gid in aGroup fix the issue

**Comment by gabrieljenik** [comment]
> I didn't reprocess any information : i use
Ok, but it requires to update all themes.

I prefer to apply central fix, out of twig, and then if someone wants to update their themes to not reprocess (although its harmless) they can.

**Comment by Shnoulle** [comment]
> I prefer to apply central fix, out of twig, and then if someone wants to update their themes to not reprocess (although its harmless) they can.

Yes, +1 , remind to apply to group name too :) It can contain GID ;)

**Comment by olleharstedt** [comment]
Are the discussions resolved?

**Comment by Shnoulle** [comment]
@maziminke : you confirm ?

**Comment by gabrieljenik** [comment]
Added a new commit to the PR as to deal with replacements on the group name

**Comment by gabrieljenik** [comment]
This is all implementes.
As soon as @maziminke confirms, this is ready for testing cycle.

**Comment by Shnoulle** [comment]
> Added a new commit to the PR as to deal with replacements on the group name

Maybe : must not send GROUPNAME inside groupname replacement ? 

(someone must test set group-name to `This is {GROUPNAME}` 😱 🤣

File: application/helpers/SurveyRuntimeHelper.php [hunk 2]
```diff
@@ -417,7 +417,7 @@ public function run($surveyid, $args)
                 $aGroup['class'] = ' ls-hidden';
             }
 
-            $aGroup['name']        = $gl['group_name'];
+            $aGroup['name']        = LimeExpressionManager::ProcessString($gl['group_name'], null, null, 3, 1);
```
Comment (c-schmitz) [inline application/helpers/SurveyRuntimeHelper.php:420]: You will need to strop {GROUPNAME} here.

Using the text " This is {GROUPNAME} inception" in the groupname leads to

"This is This is This is This is Group 2 inception inception inception inception "

and will probably create unnecessary load.

**Review by c-schmitz — CHANGES_REQUESTED** [review]

File: application/helpers/SurveyRuntimeHelper.php (hunk 2 above)
Comment (Shnoulle) [reply application/helpers/SurveyRuntimeHelper.php:420]: Yes, but it's a Survey manager issue.
Still : easy fix : send only {GID}

File: application/helpers/SurveyRuntimeHelper.php [hunk 3]
```diff
@@ -418,7 +418,7 @@ public function run($surveyid, $args)
 
             $aGroup['name']        = LimeExpressionManager::ProcessString($gl['group_name'], null, null, 3, 1);
             LimeExpressionManager::updateReplacementFields(array(
-                'GROUPNAME' => $gl['group_name'],
+                'GROUPNAME' => $aGroup['name'],
```
Comment (gabrieljenik) [inline application/helpers/SurveyRuntimeHelper.php:421]: Using the processed string as replacement value to avoid recursion.

**Pull request merged**